packages/pubsub/src/subscription.js

```diff
@@ -592,157 +444,397 @@ Subscription.prototype.decorateMessage_ = function(message) {
-    returnImmediately: !!options.returnImmediately,
-    maxMessages: options.maxResults
+    pushConfig: config
   };
 
-  this.activeRequest_ = this.parent.request(protoOpts, reqOpts, function(err) {
-    self.activeRequest_ = null;
+  this.request({
+    client: 'subscriberClient',
+    method: 'modifyPushConfig',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
 
-    var resp = arguments[1];
+/**
+ * Nacks the provided message. If the connection pool is absent, it will be
+ * placed in an internal queue and sent out after 1 second or if the pool is
+ * re-opened before the timeout hits.
+ *
+ * @private
+ *
+ * @param {object} message - The message object.
+ */
+Subscription.prototype.nack_ = function(message) {
+  this.breakLease_(message);
 
+  if (!this.connectionPool) {
+    this.inventory_.nack.push(message.ackId);
+    this.setFlushTimeout_();
+    return;
+  }
+
+  var self = this;
+
+  this.connectionPool.acquire(message.connectionId, function(err, connection) {
     if (err) {
-      if (err.code === 504) {
-        // Simulate a server timeout where no messages were received.
-        resp = {
-          receivedMessages: []
-        };
-      } else {
-        callback(err, null, resp);
-        return;
-      }
+      self.emit('error', err);
+      return;
     }
 
-    var messages = arrify(resp.receivedMessages)
-      .map(function(msg) {
-        return Subscription.formatMessage_(msg, self.encoding);
-      })
-      .map(self.decorateMessage_.bind(self));
+    connection.write({
+      modifyDeadlineAckIds: [message.ackId],
+      modifyDeadlineSeconds: [0]
+    });
+  });
+};
+
+/**
+ * Opens the ConnectionPool.
+ *
+ * @private
+ */
+Subscription.prototype.openConnection_ = function() {
+  var self = this;
+  var pool = this.connectionPool = new ConnectionPool(this);
 
-    self.refreshPausedStatus_();
+  pool.on('error', function(err) {
+    self.emit('error', err);
+  });
 
-    if (self.autoAck && messages.length !== 0) {
-      var ackIds = messages.map(prop('ackId'));
+  pool.on('message', function(message) {
+    self.emit('message', self.leaseMessage_(message));
 
-      self.ack(ackIds, function(err) {
-        callback(err, messages, resp);
-      });
-    } else {
-      callback(null, messages, resp);
+    if (self.hasMaxMessages_() && !pool.isPaused) {
+      pool.pause();
     }
   });
+
+  pool.once('connected', function() {
+    clearTimeout(self.flushTimeoutHandle_);
+    self.flushTimeoutHandle_ = null;
+    self.flushQueues_();
+  });
+};
+
+/**
+ * Modifies the ack deadline on messages that have yet to be acked. We update
+ * the ack deadline to the 99th percentile of known ack times.
+ *
+ * @private
+ */
+Subscription.prototype.renewLeases_ = function() {
+  var self = this;
+
+  this.leaseTimeoutHandle_ = null;
+
+  if (!this.inventory_.lease.length) {
+    return;
+  }
+
+  var ackIds = this.inventory_.lease;
+  this.ackDeadline = this.histogram.percentile(99);
```

Comment:
Are we overwriting ```this.ackDeadline``` here on purpose rather than computing ackDeadline in a local var? Is it a feature that the user can read the ackDeadline value through this field? Do we watn to expose this to the user?